### PR TITLE
feat(payload): make the actual payload build function generic

### PR DESCRIPTION
**Pull request by tcoratger**
Should solve #4145.

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
I think the issue wanted to have an impl of `PayloadBuilder` for `Fn(BuildArguments<Pool, Client>)`, so you could simply pass a function with that signature as the payload builder without having to impl the trait for e.g. unit structs (like we do in this PR) cc @mattsse

**Comment by tcoratger** [comment]
@onbjerg Tell me if this is better now :)

**Review by onbjerg — APPROVED** [review]
lgtm, pending @mattsse

File: crates/payload/basic/src/lib.rs [hunk 1]
```diff
@@ -535,199 +538,215 @@ enum BuildOutcome {
...
     cached_reads: CachedReads,
     config: PayloadConfig,
     cancel: Cancelled,
     best_payload: Option<Arc<BuiltPayload>>,
-    to_job: oneshot::Sender<Result<BuildOutcome, PayloadBuilderError>>,
-) where
+}
+
+trait PayloadBuilder<Pool, Client> {
+    fn try_build(
```
Comment (mattsse) [inline crates/payload/basic/src/lib.rs:551]: this needs some docs and we need to make this public

File: crates/payload/basic/src/lib.rs [hunk 2]
```diff
@@ -535,199 +538,215 @@ enum BuildOutcome {
...
+{
+    fn try_build(
+        &self,
+        args: BuildArguments<Pool, Client>,
+    ) -> Result<BuildOutcome, PayloadBuilderError> {
+        self(args)
+    }
+}
+
+fn default_payload_builder<Pool, Client>(
```
Comment (mattsse) [inline crates/payload/basic/src/lib.rs:569]: I'd also like to see some basic docs here

File: crates/payload/basic/src/lib.rs [hunk 3]
```diff
@@ -535,199 +538,215 @@ enum BuildOutcome {
     Cancelled,
 }
 
-/// Builds the payload and sends the result to the given channel.
-fn build_payload<Pool, Client>(
+struct BuildArguments<Pool, Client> {
```
Comment (mattsse) [inline crates/payload/basic/src/lib.rs:541]: this and the fields also need to be public + docs

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a great first step.

we need to expose the types and make this configurable in the `BasicPayloadJobGenerator` so it can be plugged in there.

File: crates/payload/basic/src/lib.rs (hunk 2 above)
Comment (tcoratger) [reply crates/payload/basic/src/lib.rs]: @mattsse Done

File: crates/payload/basic/src/lib.rs (hunk 3 above)
Comment (tcoratger) [reply crates/payload/basic/src/lib.rs]: @mattsse Done

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**